### ISSUE TITLE
A possible bug when creating a wrapper for setter in astropy.modeling.Parameter

### ISSUE THREAD

**Issue by mcara** [astropy/astropy#4586]
I wanted to create a setter with two parameters: value and model. However, initialization of my example model class fails (note: my model is really for testing purposes and is not intended to do anything logical) :

```python
from astropy.modeling import FittableModel
from astropy.modeling.parameters import Parameter

def mysetter(val, model):
  return val + 1 + model.d

class A(FittableModel):

  outputs = ('z',)

  x = Parameter(default=1, setter=mysetter)
  y = Parameter(default=1)

  def __init__(self, x, y, d):
    self.d = d
    super(A, self).__init__(x, y)

  def evaluate(self, x, y):
    # do something:
    return x**2+y**2
```
Then
```python
>>> a=A(1,1,2)
Traceback (innermost last):
  File "<console>", line 1, in <module>
  File "<console>", line 7, in __init__
  File ".../site-packages/astropy/modeling/core.py", line 683, in __init__
    self._initialize_constraints(kwargs)
  File ".../site-packages/astropy/modeling/core.py", line 1336, in _initialize_constraints
    param = getattr(self, param_name)
  File ".../site-packages/astropy/modeling/parameters.py", line 186, in __get__
    parameter._bind(obj)
  File ".../site-packages/astropy/modeling/parameters.py", line 597, in _bind
    self._setter = self._create_value_wrapper(self._setter, model)
  File ".../site-packages/astropy/modeling/parameters.py", line 675, in _create_value_wrapper
    model_arg = inputs.args[1].name
AttributeError: 'list' object has no attribute 'args'
```

@embray Is this a bug or am I doing something wrong?

**Comment by pllim** [comment]
@mcara, what you wrote above does not quite look like http://docs.astropy.org/en/latest/modeling/new.html#defining-new-fitter-classes .

**Comment by pllim** [comment]
Oh, please ignore my previous comment. I thought you were trying to implement a fitter.

For a model, try this example: http://docs.astropy.org/en/latest/modeling/new.html#full-example

**Comment by mcara** [comment]
@pllim Your last link does not show how to define or use a parameter setter when I need this setter to get the model too (not just the value).

**Comment by pllim** [comment]
Perhaps @eteq or @nden can clarify? But I don't need it is a bug.

**Comment by pllim** [comment]
@mcara , I am able to get something to work here by tweaking your example above:
```python
from astropy.modeling import Fittable2DModel

class A(Fittable2DModel):
    @staticmethod
    def evaluate(xc, yc):
        return xc**2 + yc**2 + 2
```

```
>>> a = A()
>>> a(1, 1)
4.0
```

**Comment by pllim** [comment]
After talking to @mcara , I am now convinced it is a bug in `astropy.modeling.parameters.Parameter._create_value_wrapper()` static method. It tries to access a `.args` attribute but `inputs` is just a regular list. Proposed fix is to change

```python
model_arg = inputs.args[1].name
```

to

```python
model_arg = inputs[1].name
```

This bug was introduced by @mdboom in b38833281f.

**Comment by pllim** [comment]
A test that I did with @mcara to illustrate this bug further:

```python
In [1]: from astropy.modeling import utils

In [2]: def mysetter(val, model):
      return val + 1
   ...: 

In [3]: inputs, dummy = utils.get_inputs_and_params(mysetter)

In [4]: inputs
Out[4]: [<Parameter at ... 'val'>, <Parameter at ... 'model'>]

In [5]: inputs[1].name
Out[5]: 'model'

In [6]: inputs[1]
Out[6]: <Parameter at ... 'model'>

In [7]: inputs.args[1].name
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-...> in <module>()
----> 1 inputs.args[1].name

AttributeError: 'list' object has no attribute 'args'
```

**Comment by mcara** [comment]
I created a pull request #4655 with a fix to this bug thanks to @pllim

**Comment by pllim** [comment]
PR is now #4656

**Comment by embray** [comment]
For what purpose are you planning on adding a "setter" for a parameter?  I had actually planned to remove that feature, or at least significantly rework it.

**Comment by pllim** [comment]
@embray , it was for PSF fitting sprint for JWST DADF. If I understood correctly from @mcara , he no longer needs this feature but it is still a bug to be fixed.

**Comment by eteq** [comment]
fixed in #4656
